### PR TITLE
Add support for PN532 based RFID readers

### PR DESCRIPTION
**Pull request by fredg02**
Fix for issue #824.

I've added the py532lib from https://github.com/HubCityLabs/py532lib. Let me know if the library should only be added if it's required (e.g. with a little install script).

This has been tested successfully with the latest source code from the develop branch on Raspbian Stretch with a Raspberry Pi Zero.
There is a potential issue where the I2C connection is lost after a while. A manual workaround is to reset the I2C connection via raspi-config. It should be easy to automate this resetting procedure if it occurs more frequently (during my short tests it was pretty stable).

If this PR gets merged, I can add a section to the wiki explaining the hard- and software setup (incl. wiring), etc.

### CODE REVIEW

**Comment by s-martin** [comment]
Thanks for your contribution to Phoniebox!

A few thoughts:
- dependencies could by installed via pip, see e.g. https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/5b2cf742573489db27b117bf9588f2a54d750365/requirements.txt#L8 so we would not need to copy the other libs into the repo
- the used Quickwire library is licensed under GPL, we need to check, how we can use it to be compliant with the license

We are happy to assist you to get this new code parts integrated here!

**Comment by s-martin** [comment]
So the Quickwire license seems to be LGPL or MIT license, maybe it's not really an issue. 

Neverthelesss we should decide, if we want to distribute the code (i.e. add it to the repo here) or just load via pip.

**Comment by fredg02** [comment]
Hi,

Thanks for the quick reply.

>  * dependencies could by installed via pip, see e.g. https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/5b2cf742573489db27b117bf9588f2a54d750365/requirements.txt#L8
>   so we would not need to copy the other libs into the repo

Unfortunately it does not work via `pip install py532lib`:

    Collecting py532lib
    Could not find a version that satisfies the requirement py532lib (from versions: )
    No matching distribution found for py532lib`
It does work with `pip3 install py532lib` though. Where and how exactly should this be installed? Also in the requirements.txt file?

>  * I have seen the py532lib hasn't been updated for a few years, so is the current state of the code ok for our use case here?

As mentioned, during my (limited) tests I have not found any issues. So I'm assuming that this legacy code is good enough. There are two alternative Python libs for PN532, but they have similar/other limitations:
* https://github.com/adafruit/Adafruit_Python_PN532 (quite old as well, repo is archived (read-only) and it seems to work with SPI only, not I2C)
* https://github.com/adafruit/Adafruit_CircuitPython_PN532 (not quite as old, but depends on CircuitPython and also works with SPI only AFAICT)

**Comment by s-martin** [comment]
pip3 is fine for us, we are in the process of moving to Python 3 only anyway.

My suggestion would be to create a file e.g. `requirements-pn532.txt` (for example see https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/362fb7742a59fc2691436319b0ef0574002073c2/requirements-spotify.txt), which installs all dependencies for that reader.

If we want to add the call `sudo python3 -m pip install -r requirements-pn532.txt` to the regular installer or we provide a README with a description how users could install it, should be discussed. @MiczFlor and @ZyanKLee , what do you think?

Generally I suggest to check  https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/CONTRIBUTING.md, there is some info provided how additional components should be added to the repo.

You may also want to check #790, what's currently discussed and maybe #792 for future handling of additional components (plugins).

**Comment by ZyanKLee** [comment]
My goal would be to assemble the pip-install command based on the user's input within our installer and install all pypi dependencies at once. That should avoid version incompatibilities... Hopefully.

**Comment by fredg02** [comment]
Based on #790 I have created `components/rfid-reader/rfid-reader-PN532/requirements-pn532.txt` and referenced it directly in the installer scripts for now.

**Side note:** mentioning the type of the component in the subsubfolder again seems redundant to me. I'm not even sure that the component name has to be added to the requirements.txt's file name.
E.g. `components/rfid-reader/PN532/requirements.txt` seems unique enough to me. I will add a comment to issue #790.

File: scripts/installscripts/stretch-install-default.sh [hunk 1]
```diff
@@ -410,6 +410,7 @@ git fetch
 
 # Install more required packages
 sudo pip install -r requirements.txt
+sudo pip3 install -r /home/pi/RPi-Jukebox-RFID/components/rfid-reader/rfid-reader-PN532/requirements-pn532.txt
```
Comment (s-martin) [inline scripts/installscripts/stretch-install-default.sh]: I would not add the installation to the scripts for stretch.

Stretch is still much Python 2 code and I would expect some problems on stretch. If anyone wants to use that reader on stretch, it's still possible to install it manually

File: scripts/installscripts/stretch-install-spotify.sh [hunk 2]
```diff
@@ -509,6 +509,7 @@ git fetch
 
 # Install more required packages
 sudo pip install -r requirements.txt
+sudo pip3 install -r /home/pi/RPi-Jukebox-RFID/components/rfid-reader/rfid-reader-PN532/requirements-pn532.txt
```
Comment (s-martin) [inline scripts/installscripts/stretch-install-spotify.sh]: see above

File: scripts/installscripts/stretch-install-default.sh (hunk 1 above)
Comment (fredg02) [reply scripts/installscripts/stretch-install-default.sh]: > Stretch is still much Python 2 code and I would expect some problems on stretch.

A few lines above (https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/scripts/installscripts/stretch-install-default.sh#L396) Python3 is installed. This seems to be inconsistent with your statement. I also tested it with Python3 on Stretch without any issues.

I think the way how additional components should be installed needs to be improved anyway, but until that is tackled, I'd recommend to make the install scripts as similar as possible and avoid having to document different behaviors.

Comment (s-martin) [reply scripts/installscripts/stretch-install-default.sh]: You're right, python3 is used in stretch as well, but there really wasn't much testing lately.

Users are encouraged to use buster for new installations (e.g. mopidy 3 for Spotify edition won't work with stretch anymore). 

So we don't really get any feedback, if new installations on stretch still work.

Nevertheless I don't see much harm, if stretch also installs this reader.

**Comment by fredg02** [comment]
I will update this PR with the shorter path to the requirements.txt as described above.

File: scripts/Reader.py.experimental [hunk 3]
```diff
@@ -125,6 +127,21 @@ class Rdm6300Reader:
...
+        self.device = Mifare()
+        self.device.SAMconfigure()
+        self.device.set_max_retries(MIFARE_WAIT_FOR_ENTRY)
+
+    def readCard(self):
+        return str(+int('0x' + self.device.scan_field().hex(),0))
+
+    def cleanup(self):
+        #TODO
+        logger.debug("PN532Reader clean up.")
```
Comment (s-martin) [inline scripts/Reader.py.experimental]: Is anything to be done in the cleanup function? Because there's a *TODO*

Comment (fredg02) [reply scripts/Reader.py.experimental]: Good catch. Honestly, I don't know if anything needs to be done here for the PN532 reader.

Comment (fredg02) [reply scripts/Reader.py.experimental]: I've removed the TODO

File: scripts/installscripts/buster-install-default.sh [hunk 4]
```diff
@@ -545,6 +545,7 @@ cd /home/pi/RPi-Jukebox-RFID || exit
 
 # Install more required packages
 sudo pip install -r requirements.txt
+sudo pip3 install -r /home/pi/RPi-Jukebox-RFID/components/rfid-reader/PN532/requirements.txt
```
Comment (ZyanKLee) [inline scripts/installscripts/buster-install-default.sh]: @s-martin @MiczFlor at some point we should think about if it is a good idea to install all dependencies for all hardware-combinations we can think of. but for now I'm fine with installing one more.

**Comment by ZyanKLee** [comment]
@fredg02 did you do a clean install with those changes?

**Comment by fredg02** [comment]
> @fredg02 did you do a clean install with those changes? And does the phoniebox with your RFID reader work after that?

Not yet, but I will test it.

**Comment by s-martin** [comment]
If you think there's a need to explain how to use this feature, please add a short `README.md` to  `components/rfid-reader/PN532/`.

File: scripts/installscripts/buster-install-default.sh (hunk 4 above)
Comment (MiczFlor) [reply scripts/installscripts/buster-install-default.sh]: I think that - for the time being - we won't clutter the SD card of the Pi with the libraries. The install process will take a little longer, the response time should make no difference.
So I agree for now.

**Comment by MiczFlor** [comment]
@fredg02 thanks for your contribution. And very happy to see that the `components` folder is coming to life.

There is a conflict with the two install scripts for buster and stretch at the moment. Could you have a look please and fix this? It seems that later changes compromised your contribution.

And if you add a few lines to a `README.md` in the folder where the requirements are, I would later add that to the wiki

Thanks! micz

**Comment by fredg02** [comment]
Yes, I will. Sorry for the delay. Priorities have been shifting around lately, so I did not get to work on the PR.

**Comment by fredg02** [comment]
Conflicts in the install scripts have been fixed. README.md has been added.

I've successfully tested the install script for Buster on a clean Raspbian Buster installation. After the installation I had to follow the steps in the README.md to get the PN532 reader to work.

File: components/rfid-reader/PN532/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+The following steps should be done before installing RPi-Jukebox-RFID. If RPi-Jukebox-RFIS is already
+installed, make sure to disable the phoniebox-rfid-reader service first (`sudo systemctl stop phoniebox-rfid-reader.service`).
+   
+
+1. Connect the PN532 RFID reader to the GPIO pins
+
+    | PN532 | Raspberry |
+    | ----- | --------- |
+    | 5V    | 5V        |
```
Comment (s-martin) [inline components/rfid-reader/PN532/README.md]: Very minor improvement: for convenience you could add the pin number on the Pi.

Comment (fredg02) [reply components/rfid-reader/PN532/README.md]: I've added the pin numbers.

**Comment by fredg02** [comment]
@s-martin, @ZyanKLee and @MiczFlor:Thanks for your support with this PR.

**Pull request merged**